### PR TITLE
client-go/transport: drop Dial and GetCert fields in favor of Holders

### LINKED ISSUE #112342: Remove client-go TLS cache dual function/holder fields
These were introduced in #112017 for the sake of backwards compatibility and should be removed.

/kind cleanup
/assign
/milestone v1.26

### PR DESCRIPTION
**Pull request by enj**
Signed-off-by: Monis Khan <mok@microsoft.com>

/kind cleanup
/assign @liggitt

Fixes #112342

```release-note
NONE
```

### CODE REVIEW

**Comment by enj** [comment]
cc @aojea

**Comment by aojea** [comment]
I think this deserves release-note, but I'll wait for Jordan's opinion on deprecating those fields

**Comment by enj** [comment]
> I think this deserves release-note, but I'll wait for Jordan's opinion on deprecating those fields

I think release notes are targeted at users of the Kubernetes binaries and not the code libraries (which have no backwards compatibility requirements).  As one data point, I have used client-go for years and I have never read the release notes for anything related to them.

**Comment by aojea** [comment]
> I think release notes are targeted at users of the Kubernetes binaries and not the code libraries (which have no backwards compatibility requirements). As one data point, I have used client-go for years and I have never read the release notes for anything related to them.

fair enough to me :+1:

**Comment by aojea** [comment]
/lgtm

_Less is more_

I personally think that release-notes should be for everything that the projects exposes to users.
There are a lot of consumers of client-go, and if somebody breaks with this changes we can always point them to the release notes ... but this is totally personal opinion, I'm completely ok with your rationale too

File: staging/src/k8s.io/client-go/transport/cache.go [hunk 1]
```diff
@@ -93,13 +93,13 @@ func (c *tlsTransportCache) get(config *Config) (http.RoundTripper, error) {
...
-	if tlsConfig == nil && config.Dial == nil && config.Proxy == nil {
+	if tlsConfig == nil && config.DialHolder == nil && config.Proxy == nil {
 		return http.DefaultTransport, nil
 	}
 
 	var dial func(ctx context.Context, network, address string) (net.Conn, error)
-	if config.Dial != nil {
-		dial = config.Dial
+	if config.DialHolder != nil {
+		dial = config.DialHolder.Dial
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/transport/cache.go:102]: can we rely on DialHolder.Dial being non-nil here?

**Review by liggitt — COMMENTED** [review]
a couple questions, lgtm otherwise

File: staging/src/k8s.io/client-go/transport/cache.go (hunk 1 above)
Comment (enj) [reply staging/src/k8s.io/client-go/transport/cache.go:102]: Yes, this one is safe.

File: staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go [hunk 2]
```diff
@@ -308,17 +308,14 @@ func (a *Authenticator) UpdateTransportConfig(c *transport.Config) error {
 	if c.HasCertCallback() {
 		return errors.New("can't add TLS certificate callback: transport.Config.TLS.GetCert already set")
 	}
-	c.TLS.GetCert = a.getCert.GetCert
 	c.TLS.GetCertHolder = a.getCert // comparable for TLS config caching
 
-	if c.Dial != nil {
+	if c.DialHolder != nil {
 		// if c has a custom dialer, we have to wrap it
 		// TLS config caching is not supported for this config
```
[the remark this answers is not in the document]
Comment (enj) [reply staging/src/k8s.io/client-go/plugin/pkg/client/auth/exec/exec.go:315]: This one is not safe, let me fix it.

**Comment by enj** [comment]
@liggitt should be good now.

**Pull request merged**